clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+filtered_gdf.to_csv(os.path.join(cache_dir, 'HLSL30_filtered.csv'), index=False)
+
+# %%
+for granule in granules:
+    fpath = os.path.join(cache_dir, granule['umm']['GranuleUR'])
+    if not os.path.exists(fpath):
+        earthaccess.download([granule], fpath)
+
+# %% [markdown]
+# ### Open, crop, and mask data
+# 
+# This will be the most resource-intensive step. I recommend caching your results using the `cached` decorator or by writing your own caching code. I also recommend testing this step with one or two dates before running the full computation.
+# 
+# This code should include at least one **function** including a numpy-style docstring. A good place to start would be a function for opening a single masked raster, applying the appropriate scale parameter, and cropping.
+# 
+# **YOUR TASK:**
+# 1. For each granule:
+#     1. Open the Fmask band, crop, and compute a quality mask for the granule. You can use the following code as a starting point, making sure that `mask_bits` contains the quality bits you want to consider:
+#          ```python
+#          # Expand into a new dimension of binary bits
+#          bits = np.unpackbits(da.astype(np.uint8)).reshape(da.shape + (-1,))
+# 
+#          # Select the required bits and check if any are flagged
+#          mask = np.prod(bits[..., mask_bits]==0, axis=-1)
+#          ```
+# 
+#     2. For each band that starts with 'B':
+#         1. Open the band, crop, and apply the scale factor
+#         2. Name the DataArray after the band using the `.name` attribute
+#         3. Apply the cloud mask using the `.where()` method
+#         4. Store the DataArray in your data structure (e.g. adding a GeoDataFrame column with the DataArray in it. Note that you will need to remove the rows for unused bands)
+
+# %%
+def decode_qa(bit_string):
+    """
+    Decode an 8-bit string into its quality assessment components.
+
+    Bits are listed from the MSB (bit 7) to the LSB (bit 0): 
+        7-6    aerosol:
+            00 - climatology
+            01 - low
+            10 - average
+            11 - high
+        5      water
+        4      snow/ice
+        3      cloud shadow
+        2      adjacent to cloud
+        1      cloud
+        0      cirrus cloud
+    
+    Parameters:
+    bit_string (str): An 8-bit string representing quality assessment flags.
+    
+    Returns:
+    dict: A dictionary with human-readable descriptions of the QA flags.
+    """
+    # Define the bit positions based on the provided table
+    aerosol_level_bits = bit_string[:2]
+    water_bit = bit_string[2]
+    snow_ice_bit = bit_string[3]
+    cloud_shadow_bit = bit_string[4]
+    adjacent_cloud_shadow_bit = bit_string[5]
+    cloud_bit = bit_string[6]
+    
+    # Map the bit values to their meanings
+    aerosol_levels = {
+        '00': 'Climatology aerosol',
+        '01': 'Low aerosol',
+        '10': 'Moderate aerosol',
+        '11': 'High aerosol'
+    }
+    
+    # Decode the bits using the mapping
```

Comment:
Yeah, I agree with @peterkobylarz here, really nice code -- obviously my way was a bit error-prone since I got my bit order reversed! I suspect, however, that my way is also quite a bit faster, so working with larger datasets it might be worthwhile to write an alternative that uses vectorized numpy methods, perhaps tested against this very clear code. Decoding bits could also be a great application of object-oriented programming in Python, especially as this is a common operation (but what the bits decode to is not standard).